test/puppetlabs/puppetdb/http/command_test.clj

```diff
@@ -139,3 +136,110 @@
         (let [timestamp (get-in good-msg [:headers :received])]
           (time/parse (time/formatters :date-time) timestamp))))))
 
+(deftest normalize-request
+  (let [normalize (#'tgt/normalize-request identity)]
+
+    ;; Check for an error if some but not all params are present
+    (doseq [items (apply concat
+                         (map #(combinations ["command" "version" "certname"] %)
+                              [1 2]))]
+      (let [result (normalize {:body ::ignored
+                               :params (into {} (for [k items] [k "1"]))})]
+        (is (= 400 (:status result)))
+        (is (= (apply str
```

Comment:
I look at it as a tradeoff (in the absence of more discriminating exceptions), if you're not specific enough, then you of course risk false negatives, and false positives mean you have to update the tests with the code, which doesn't bother me too much as long as it's not too frequent.